Prevent <<command("parameters",123)>> from crashing bondage.js

blurymind opened the pull request:
Gave this another shot - this time from a different angle. Much thanks to @jhayley for the kind feedback and patience :) 

There is still a quirk seen here:
https://github.com/jhayley/bondage.js/pull/40/files#diff-1f6fe15287b398db63504fd4a8ee9bbbR404
Not sure why bondage.js adds an extra space before the "("

jhayley commented:
This is looking good so far! The space thing is an issue though, I think it's related to this line https://github.com/jhayley/bondage.js/blob/40748d2e50e26f40398df0a01e4d0bc070fe35e6/src/parser/parser.js#L116

blurymind commented:
Ah yes, I think you're right. I had my suspicions, but now its pretty clear. I will try to tackle this tonight :)

jhayley commented on src/lexer/tokens.js:
```diff
@@ -87,10 +87,10 @@ const Tokens = {
   Comment:              '//',                      // a run of text that we ignore
 
-  Identifier:           /[a-zA-Z0-9_:.]+/,         // a single word (used for functions)
+  Identifier:           /[ a-zA-Z0-9_:.]+/,         // words (used for functions)
```
Unfortunately this is not an acceptable change. an identifier is equivalent to a variable name in a programming language, so it shouldn't have spaces in it. The fact that this didn't break any tests means the tests are lacking lol, whoops.
We'll need to find another way

blurymind replied:
ah ok, I thought that its a bit hacky

blurymind replied:
Ok you got me there xD I have no idea how to remove the ugly hack, without breaking the parser tests. I get this error:

  1) Dialogue Returns a command to the user:
     Error: Parse error on line 2: Unexpected 'Text'
      at Parser.parseError (eval at createParser (node_modules\jison\lib\jison.js:1327:13), <anonymous>:156:21)
      at Parser.parse (eval at createParser (node_modules\jison\lib\jison.js:1327:13), <anonymous>:223:22)
      at Runner.run (src\runner.js:84:43)
      at run.next (<anonymous>)
      at Context.it (tests\test_runner.js:397:21)

When I take it out

blurymind replied:
If I add () to the identifier token, that breaks a whole bunch of other tests.. 
Whatever angle I try with it, it breaks

blurymind replied:
It works if I create a new token for commandIdentifier , but we still need to keep the ugly hack.

blurymind replied:
Ok no need to use the ugly hack any more and also fixed this without changing the Identifier's regex

blurymind commented:
@jhayley I fixed the problem with the inserted space, but will need to keep the ugly parser 'hack' (see previous comments on why)

blurymind commented:
@jhayley After ranting with myself for a bit here, I eventually created a new 'CommandCall' token and removed the ugly hack altogether. All tests passing, simplified command parser code and anything allowed inside << >>, even spaces - without using a workaround or artificial insertion of spaces

jhayley commented:
Looks good! Will push to npm when I have a chance

jhayley commented:
pushed! ✨
